### ISSUE TITLE
Slow debugger response

### ISSUE THREAD

**Issue by mmorhun** [eclipse-che/che#7156]
### Description
Sometimes after debugger stops on a breakpoint it takes more than 10 seconds while frames and local variables displayed.

### Reproduction Steps
1. Open Che in Che (che-dev image)
2. Start another Che instance
3. Open a elenium test which uses `importJavaApp` method and make url wrong
4. From host OS start the test and stop execution with breakpoint on the method.
4. Open `Network` tab in browser and do `Step over`

When debugger stops on exception throwing, contentent of debugger panel will be delayed for relatively very long (~20 seconds in my case) time.

**Screencast:**
![slow-debugger-response](https://user-images.githubusercontent.com/15607393/32329947-2c27a4f4-bfe7-11e7-8d4d-0f71396d0aed.gif)

**OS and version:**  Fedora 25, Che 6.0.0-M2, mounted Che project.

**Comment by mmorhun** [comment]
it is reproduced just with debugging of `@BeforeClass` method. First stop on breakpoint, put another one in the method then resume.
Result:
![screenshot from 2017-11-02 16-43-27](https://user-images.githubusercontent.com/15607393/32332114-51e3a23c-bfed-11e7-8c0e-61b23f1e30bc.png)

**Comment by tolusha** [comment]
It is needed to profile server-side part
https://github.com/codenvy/codenvy/issues/1466

**Comment by mmorhun** [comment]
I figured out, that `org.eclipse.jdt.core.search.SearchEngine.searchAllTypeNames` tooks most of the time (~78 seconds)

**Comment by tsmaeder** [comment]
@mmorhun is that in the che6 branch?

**Comment by mmorhun** [comment]
Unfortunately, yes

**Comment by mmorhun** [comment]
@tsmaeder it is reproduced for workspaces with huge project only (like Che)

**Comment by tolusha** [comment]
![screenshot from 2017-12-11 12-34-09](https://user-images.githubusercontent.com/1640675/33827045-9ee8e79a-de6f-11e7-9371-496ee5e7b962.png)

**Comment by tsmaeder** [comment]
@tolusha do we know how many times the search is called? Maybe we do it too often?

**Comment by tolusha** [comment]
I think so too. I will dig more

**Comment by mmorhun** [comment]
Short instruction how to profile wsagent with VisualVM:
1. Create stack for workspace from recipe:
```
FROM eclipse/ubuntu_jdk8

USER root
RUN echo "user:password" | chpasswd
RUN echo 'X11UseLocalhost yes' >> /etc/ssh/sshd_config && echo 'AddressFamily inet' >> /etc/ssh/sshd_config

USER user
RUN cd /home/user && wget -q https://github.com/visualvm/visualvm.src/releases/download/1.4/visualvm_14.zip && unzip visualvm_14.zip && rm visualvm_14.zip
```
And add server for ssh (22) port on stack creation step (or add from dashboard later)

2. Create and run workspace from the stack. Retrieve port on which ssh (22) port of the workspace was exposed. (using `docker ps` or Che API for example)

3. From host run `ssh -X -p <port> user@127.0.0.1 /home/user/visualvm_14/bin/visualvm` where `<port>` should be replaced with obtained port from step `2`. Default password is `password`, but it can be easily changed in the recipe.

4. Now VisualVM window should appear on the host. Connect to tomcat from the `local` list. 

Note:
- You don't have to have VisualVM installed on host.
- This is tested for local Che instance, but I believe it works for remote, just ip/hostname should be adjusted.

**Comment by mmorhun** [comment]
Profiling result:
![prof](https://user-images.githubusercontent.com/15607393/34378808-90556d46-eb00-11e7-8295-84c1dea77842.png)

**Comment by mmorhun** [comment]
Depends on: https://github.com/eclipse/eclipse.jdt.ls/issues/510

**Comment by gorkem** [comment]
As commented on the jdt.ls issue, I do not think `SearchEngine` will perform predictably for this kind of use especially on large projects.
 
@mmorhun You can take a look at https://github.com/eclipse/eclipse.jdt.debug/tree/5ae085d07ebb148b424b5ea934d46101e1adbb1f/org.eclipse.jdt.launching/launching/org/eclipse/jdt/launching/sourcelookup for inspiration on how to implement source lookup.  Basically, it is about iterating over all the source locations (source folders, archives with source, libraries) and look for entries that match the FQN.

**Comment by mmorhun** [comment]
I replaced `SearchEngine` in JDT LS branch (_see links above_) and did profiling on our [spring sample](https://github.com/che-samples/web-java-spring) (breakpoints on the first and the last line of the `GreetingController`). Here is the results:

| Fix | Sample 1 |  Sample 2 |  Sample 3 |  Sample 4 |  Average |
| ------------ | ------------ | ------------- | ------------- | ------------- | ------------- |
| Before, `ms`  | 622 | 865 | 717 | 709  | 728.25 |
| After,    `ms`  | 132 | 136 | 153 | 138  | 139.75 |

This is about `5.2` times faster.

<details>
<summary>Screenshots</summary>
Before:
 
![before](https://user-images.githubusercontent.com/15607393/35507962-3e52a386-04f7-11e8-8a71-d92da588c04c.png)

After: 
![after](https://user-images.githubusercontent.com/15607393/35507970-4b324e4e-04f7-11e8-85b9-3c8927d4492a.png)

</details>

**Comment by tolusha** [comment]
It is deserved to be ported to eclipse/che master branch.

**Comment by mmorhun** [comment]
I ported the solution to master, but it turned out that we have yet another problem which doesn't depend on `SearchEngine`.
I noticed, that thread dump request take about 3-4 seconds in a 'fresh' workspace and going slower and slower with time. I made an experiment: started Che, created new workspace, opened a file and just left it until next day. Then I tried to debug selenium test. Thread dump requests duration were about 24-25 seconds each. But actually two hours are enough to reproduce the problem.
Also I tried to do profiling, but the results were quite different (even with workspace restarting), so I cannot make adequate conclusions for now.
